PIL/Image.py

```diff
@@ -570,18 +603,21 @@ def close(self):
 
```

Comment:
This line is erroring out with: 
```PIL/Image.py:606: error: Incompatible types in assignment (expression has type "deferred_error", variable has type "Optional[ImagingCore]")```

It's probably fixable, as we can type that as a Union, but there's also:
 
```PIL/Image.py:80: error: Incompatible types in assignment (expression has type "_imaging_not_installed", variable has type Module)```

These are both cases where we're explicitly assigning in an object that returns verbose errors if accessed. I'm not sure how a type system deals with intentional type breakage like this.

Reply:
Looking over the codebase, I think we could fix this by simply operating on `self.im` within a type's context, so that it never changes type (similar to using a `Maybe` monad in Haskell). It would greatly simplify the type signature so we wouldn't have to use `Optional[ImagingCore,deferred_error]`.

Reply:
I understand most of those words, but... I'm not sure how to apply it.